pygen/pygen_src/riscv_directed_instr_lib.py

```diff
@@ -161,3 +163,134 @@ def post_randomize(self):
+        if allow_branch:
+            # TODO `DV_CHECK_STD_RANDOMIZE_FATAL(enable_branch)
+            pass
+        else:
+            self.enable_branch = 0
+
+        if self.enable_branch:
+            self.branch_instr = \
+                riscv_instr_ins.get_rand_instr(include_category=[riscv_instr_name_t.BRANCH.name])
+            # TODO `DV_CHECK_STD_RANDOMIZE_FATAL(branch_instr)
+            self.branch_instr.imm_str = self.push_start_label
+            self.branch_instr.brach_assigned = 1
+            self.push_stack_instr[0].label = self.push_start_label
+            self.push_stack_instr[0].has_label = 1
+            self.branch_instr.extend(self.push_stack_instr)
+        self.mix_instr_stream(self.push_stack_instr)
+        for i in range(len(self.instr_list)):
+            self.instr_list[i].atomic = 1
+            if self.instr_list[i].label == '':
+                self.instr_list[i].has_label = 0
+
+
+# Pop stack instruction stream
+class riscv_pop_stack_instr(riscv_rand_instr_stream):
+
+    def __init__(self):
+        super().__init__()
+        self.stack_len = 0
+        self.num_of_reg_to_save = 0
+        self.num_of_redudant_instr = 0
+        self.pop_stack_instr = []
+        self.saved_regs = []
+
+    def init(self):
+        self.reserved_rd = [cfg.ra]
+        self.num_of_reg_to_save = len(self.saved_regs)
+        if self.num_of_reg_to_save * 4 > self.stack_len:
+            logging.critical('stack len [%0d] is not enough to store %d regs',
+                             self.stack_len, self.num_of_reg_to_save)
+        self.num_of_redudant_instr = random.randrange(3, 10)
+        self.initialize_instr_list(self.num_of_redudant_instr)
+
+    def gen_pop_stack_instr(self, stack_len, saved_regs):
+        self.stack_len = stack_len
+        self.saved_regs = saved_regs
+        self.init()
+        self.gen_instr(1)
+        self.pop_stack_instr = [None] * (self.num_of_reg_to_save + 1)
+        for i in range(len(self.pop_stack_instr)):
+            self.pop_stack_instr[i] = riscv_instr()
+
+        for i in range(len(self.saved_regs)):
+            if rcs.XLEN == 32:
+                self.pop_stack_instr[i] = riscv_instr_ins.get_instr(riscv_instr_name_t.LW.name)
+                with self.pop_stack_instr[i].randomize_with() as it:
+                    self.rd == self.saved_regs[i]
+                    self.rs1 == cfg.sp
+                    self.imm == 4 * (i + 1)
+            else:
+                self.pop_stack_instr[i] = riscv_instr_ins.get_instr(riscv_instr_name_t.LD.name)
+                with self.pop_stack_instr[i].randomize_with() as it:
+                    self.rd == self.saved_regs[i]
+                    self.rs1 == cfg.sp
+                    self.imm == 8 * (i + 1)
+
+            self.pop_stack_instr[i].process_load_store = 0
+
+        # addi sp,sp,imm
+        self.pop_stack_instr[self.num_of_reg_to_save] = riscv_instr_ins.get_instr(
+            riscv_instr_name_t.ADDI.name)
```

Comment:
I think there should be a `DV_CHECK_RANDOMIZE` call here - https://github.com/google/riscv-dv/blob/master/src/riscv_directed_instr_lib.sv#L407

Is this a todo item?

Reply:
Hi @udinator 
Yes, It's a TODO Item. Will add TODO here.